dev/com.ibm.tx.jta/src/com/ibm/tx/jta/impl/TransactionImpl.java

```diff
@@ -271,18 +273,34 @@ public void prepare() {
         }
     };
 
-    private static NewTransactionCheckpointHook getNewTxHook() {
+    private static NewTransactionCheckpointHook getNewTxCheckpointHook() {
         return NewTransactionCheckpointHook._instance;
     }
 
+    private static void addNewTxCheckpointHook() {
+        if (CheckpointPhase.getPhase().restored()) {
+            // Normal startup; do nothing
+        } else {
+            // Log the stack trace of the thread that created the transaction
+            final StringWriter writer = new StringWriter();
+            final PrintWriter printWriter = new PrintWriter(writer);
+            printWriter.println();
+            for (StackTraceElement element : Thread.currentThread().getStackTrace())
+                printWriter.println("\t" + element);
+            Tr.warning(tc, "WTRN0155_CHECKPOINT_NEW_TX_STACK", writer.getBuffer());
+
+            CheckpointPhase.getPhase().addMultiThreadedHook(getNewTxCheckpointHook());
```

Comment:
Should use an AtomicBoolean to gate the hook only being added once.

Reply:
Added the AtomicBoolean as directed.